### PR TITLE
HPCC-11162 Improve spin block performance when contended

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly mark.kelly@lexisnexis.com

### CODE REVIEW

**Comment by mckellyln** [comment]
@ghalliday this is an update on top of your latest to help with medium contended case.

File: system/jlib/jmutex.hpp [hunk 1]
```diff
@@ -411,7 +413,17 @@ class jlib_decl  SpinLock
             return;
         }
         while (unlikely(!atomic_acquire(&value)))
-            ThreadYield();
+        {
+            unsigned i = 0;
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp]: I wonder if this code should be in a non-inline function.  That will reduce the size of the generated code, and possibly increase the number of registers available to the place this is inlined.

Comment (ghalliday) [inline system/jlib/jmutex.hpp]: I think definitely - I just looked at the generated code, and it has unrolled the loop, so each use is about 100bytes bigger.  That could be painful.

A global function like

void jlib_decl spinUntilReady(atomic_t & value);

should work.

**Comment by ghalliday** [comment]
@mckellyln see comments

**Comment by mckellyln** [comment]
ok, good idea I will make changes.

**Comment by mckellyln** [comment]
@ghalliday code updated.

**Comment by ghalliday** [comment]
Out of interest, @mckellyln have you run your performance tests on this?

@richardkchapman I think this is good to merge, but please check it.

**Comment by mckellyln** [comment]
Yes, no differences for 1 thread, small increase for 6 threads, tiny increase for 32 threads.  But still significantly better than without read loop.

**Pull request merged**